library/src/main/java/org/apache/apex/malhar/lib/dedup/DeduperTimeBasedPOJOImpl.java

```diff
@@ -0,0 +1,161 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.lib.dedup;
+
+import javax.validation.constraints.NotNull;
+
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+import org.apache.apex.malhar.lib.state.managed.TimeBucketAssigner;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.api.Context.PortContext;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.Operator.ActivationListener;
+import com.datatorrent.api.annotation.InputPortFieldAnnotation;
+import com.datatorrent.lib.util.PojoUtils;
+import com.datatorrent.lib.util.PojoUtils.Getter;
+import com.datatorrent.netlet.util.Slice;
+
+public class DeduperTimeBasedPOJOImpl extends AbstractDeduper<Object> implements ActivationListener<Context>
+{
+
+  @NotNull
+  private String keyExpression;
+
+  @NotNull
+  private String timeExpression;
+
+  private long bucketSpan;
+
+  private long expireBefore;
+
+  private long referenceInstant;
+
+  private transient Class<?> pojoClass;
+
+  private transient Getter<Object, Long> timeGetter;
+
+  private transient Getter<Object, Object> keyGetter;
+
+  @InputPortFieldAnnotation(schemaRequired = true)
+  public final transient DefaultInputPort<Object> input = new DefaultInputPort<Object>()
+  {
+    @Override
+    public void setup(PortContext context)
+    {
+      pojoClass = context.getAttributes().get(PortContext.TUPLE_CLASS);
+    }
+
+    @Override
+    public void process(Object tuple)
+    {
+      processTuple(tuple);
+    }
+  };
+
+  @Override
+  protected long getTime(Object tuple)
+  {
+    return timeGetter.get(tuple);
+  }
+
+  @Override
+  protected Slice getKey(Object tuple)
+  {
+    Object key = keyGetter.get(tuple);
+    return new Slice(key.toString().getBytes());
+  }
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    TimeBucketAssigner tba = new TimeBucketAssigner();
+    tba.setBucketSpan(Duration.standardSeconds(bucketSpan));
+    tba.setExpireBefore(Duration.standardSeconds(expireBefore));
+    tba.setReferenceInstant(new Instant(referenceInstant));
+    managedState.setTimeBucketAssigner(tba);
+    super.setup(context);
+  }
+
+  @Override
+  public void activate(Context context)
+  {
+    timeGetter = PojoUtils.createGetter(pojoClass, timeExpression, Long.class);
+    keyGetter = PojoUtils.createGetter(pojoClass, keyExpression, Object.class);
+  }
+
+  @Override
+  public void deactivate()
+  {
+  }
+
+  public String getKeyExpression()
+  {
+    return keyExpression;
+  }
+
+  public void setKeyExpression(String keyExpression)
```

Comment:
Can you please add javadoc for all the setter methods? This is often the most important information that users want to see while using any operator.

Also while adding javadoc can you please add UI metatags?